ENH: Add SemiMonthEnd and SemiMonthBegin offsets #1543

adrienemery opened the pull request:
- [x] closes #1543
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

adrienemery committed:
ENH: Add SemiMonthEnd and SemiMonthBegin offsets #1543

Closes #1543

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1161,92 @@ def onOffset(self, dt):
     _prefix = 'MS'
 
 
+class SemiMonthEnd(SingleConstructorOffset):
```
make a common class which this inherits to put common routines (SemiMonthOffset)

jreback replied:
I think you can pull out a lot of the ``.apply`` method and put it there

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1161,92 @@ def onOffset(self, dt):
+        _, days_in_month = tslib.monthrange(dt.year, dt.month)
+        return dt.day == days_in_month or dt.day == 15
+
+    _prefix = 'SM'
```
move to top of class

jreback commented:
cc @sinhrks

adrienemery commented:
@jreback I have made the requested changes - I'll squash the new commit(s) once is approved.

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1161,104 @@ def onOffset(self, dt):
+        _, days_in_month = tslib.monthrange(dt.year, dt.month)
+        return dt.day in (15, days_in_month)
+
+    def _rollback_from_15(self, other):
```
you can move all of the rollforward/back to the parent class to avoid duplication I believe

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1161,104 @@ def onOffset(self, dt):
+
+    def _rollforward_from_15(self, other):
+        """Return the next offset if other is on the 15th of the month"""
+        raise NotImplementedError
```
raise AbstractMethodError instead here (same idea but more descriptive)

adrienemery commented:
@jreback Any additional changes you see?

sinhrks commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,104 @@ def onOffset(self, dt):
+        if n >= 0:
+            if n == 0 and not self.onOffset(other):
+                n = 1  # rollforward
+            for _ in range(n):
```
This looks inefficient. Can't you add ``n // 2`` months and ``n % 2`` 15 days?

adrienemery replied:
Good call - that should also simplify things.

jreback commented:
there are some benchmarks for ``.apply`` in ``asv_bench/benchmarks/timeseries``, pls add this new freq there as well.

adrienemery commented:
Will do.

adrienemery commented:
Added benchmarks and optimized `.apply` to not use inefficient looping.

benchmark results:
```bash

[  0.00%] · For pandas commit hash e7d8257c:
[  0.00%] ·· Building for conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.......................................
[  0.00%] ·· Benchmarking conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
...
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_apply   69.11μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_decr    77.51μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_decr_n  90.41μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_incr    79.20μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_incr_n  99.60μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_apply     76.50μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_decr      125.04μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_decr_n    113.05μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_incr      81.48μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_incr_n    95.23μs
```

jreback commented:
this looks good. can you see if you can add an ``apply_index`` (high-perf apply).

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -51,6 +51,43 @@ New behaviour:
+Semi-Month Offsets
+^^^^^^^^^^^^^^^^^^
+
+The ``SemiMonthEnd`` ('SM') ``SemiMonthBegin`` ('SMS') offsets provide date offsets
```
Pandas has gained new frequency offsets, ``SemiMonthEnd`` ('SM') and ``SemiMonthBegin`` ('SMS'). These provide date offsets ....

MaximilianR commented:
I am probably missing something obvious, but should the `SemiMonthBegin` be the 1st & 16th of every month? That way, a 30 day month has an equal number of days in each semi-month?

(I'm ignoring February - but maybe that should be a day earlier...)

jreback commented:
I think this is pretty standard to do semi-montly on 1/15th. @adrienemery thoughts?

adrienemery commented:
@MaximilianR @jreback I've seen both personally but how about we make it flexible such that the 15th becomes a parameter. That way people can use it for their application as needed. I like keeping 15th as the default since it provides a nice default symmetry.

See latests commit for proposed change - I haven't added tests yet, but if we agree to go this route I will, otherwise will rollback that change.

I did test it in a notebook and it works nicely:

<img width="777" alt="screenshot 2016-06-01 13 45 42" src="https://cloud.githubusercontent.com/assets/4293818/15725121/3042a8c6-27ff-11e6-8fa1-38a49078a333.png">

adrienemery commented:
I could also update `_from_name` such that it allows setting pivot day such as `SMS-16` 

```python
@classmethod
    def _from_name(cls, suffix=None):
        if not suffix:
            pivot = 15
        else:
            pivot = int(suffix)
        return cls(pivot=pivot)
```

<img width="734" alt="screenshot 2016-06-01 13 54 34" src="https://cloud.githubusercontent.com/assets/4293818/15725367/678cafc4-2800-11e6-9bdf-a5c288c94700.png">

jreback commented:
yeah that's prob reasonable though need to add routines to parse SMS-16 then

look at the last week freq to see what the parameters are named (eg not pivot)

adrienemery commented:
> look at the last week freq to see what the parameters are named (eg not pivot)

 I don't understand what you mean by this.

jreback commented:
you are thinking of allowing ``SMS-16``, so needs to parse this

```
In [2]: pd.tseries.frequencies.to_offset('1MS')
Out[2]: <MonthBegin>

In [3]: pd.tseries.frequencies.to_offset('2M')
Out[3]: <2 * MonthEnds>
```

adrienemery commented:
- added `.apply_index` methods
- added flexible `day_of_month` parameter that defaults to the 15th
- added support to parse `day_of_month` from frequency string

I'd like to add a few more tests for `.apply_index` but wanted to get feedback on current change in case there are any issues/changes to be made.

Updated asv benchmark output with `.apply_index` included:
```bash
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_apply           58.91μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_apply_index     49.61ms
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_decr            68.74μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_decr_n          80.07μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_decr_rng        70.12ms
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_incr            75.89μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_incr_n          73.19μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_begin_incr_rng        56.28ms
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_apply             73.79μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_apply_index       57.93ms
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_decr              77.41μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_decr_n            83.09μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_decr_rng          56.51ms
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_incr              66.83μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_incr_n            70.97μs
[  0.01%] ··· Running timeseries.timeseries_semi_month_offset.time_semi_month_end_incr_rng          55.86ms
```

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -51,6 +51,70 @@ New behaviour:
+Pandas has gained new frequency offsets, ``SemiMonthEnd`` ('SM') and ``SemiMonthBegin`` ('SMS').
+These provide date offsets anchored (by default) to the 15th and end of month, and 15th and 1st of month respectively.
+Optionally you can specify the ``day_of_month`` which will be used instead of the 15th each month. Anchoring suffixes
+are also supported for setting the `day_of_month``. Note that ``day_of_month`` must be 1<=day_of_month<=27 for
```
you don't need this last sentence (the Note). That can be part of the doc-string.

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -51,6 +51,70 @@ New behaviour:
+.. ipython:: python
+
+    Timestamp('2016-01-01') + SemiMonthEnd()
+
```
maybe shorten the examples and put 2 (or 3) for each sub-section.

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+        # determine how many days away from the 1st of the month we are
+        days_from_start = i.to_perioddelta('M').asi8
+        delta = Timedelta(days=self.day_of_month - 1).asm8.astype(int)
+        # get boolean array for each element before the day_of_month
```
put blank lines in between the sections (comment & code)

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -3079,6 +3079,42 @@ def test_datetime64_with_DateOffset(self):
             assert_func(result, exp)
             assert_func(result2, exp)
 
+            s = klass([Timestamp('2000-01-15 00:15:00', tz='US/Central'),
```
can you move this function into a separate test class (its getting too long), then you can also have simpler names for the class. I think further move this to ``test_offsets.py``.

jreback commented:
@chris-b1 can you have a look.

chris-b1 commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+    def apply_index(self, i):
+        # determine how many days away from the 1st of the month we are
+        days_from_start = i.to_perioddelta('M').asi8
+        delta = Timedelta(days=self.day_of_month - 1).asm8.astype(int)
```
I'm guessing this fails on windows (will try to cast to int32), use `Timedelta(...).value`

chris-b1 commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+    _min_day_of_month = 2
+
+    def __init__(self, n=1, normalize=False, **kwds):
+        self.day_of_month = kwds.get('day_of_month',
```
I know a lot of the other offsets just use `**kwds`, but where you're defining a constructor, I think it would make sense to add `day_of_month` as an argument?

adrienemery replied:
agreed

adrienemery replied:
However, as far as I can tell, to properly create Offsets from frequency strings with suffixes such as `SMS-16` you need to use `**kwds`

jreback replied:
yeah, I think you can simply include this in the argument list, or does it not work when you do?

chris-b1 commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+    def _apply_index(self, i, before_day_of_month, after_day_of_month):
+        n = self.n
+        time = i.to_perioddelta('D')
+        is_month_end = tslib.get_start_end_field(i.asi8, 'is_month_end')
```
not a big deal, but why not just `i.is_month_end`

chris-b1 commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -3115,6 +3151,7 @@ def test_datetime64_with_DateOffset(self):
             # assert these are equal on a piecewise basis
             offsets = ['YearBegin', ('YearBegin', {'month': 5}), 'YearEnd',
                        ('YearEnd', {'month': 5}), 'MonthBegin', 'MonthEnd',
+                       'SemiMonthEnd', 'SemiMonthBegin',
```
in the list of times to test (`s = klass([Timestamp() ...])` a few lines up), add a date that lands on the semi-month offset, e.g. `2000-01-15`

chris-b1 commented:
few comments, looks good overall

adrienemery commented:
I have addressed comments cc/ @jreback @chris-b1

chris-b1 commented:
Looks good to me.  If you ever hit the wall on vectorized performance, it probably wouldn't be hard to modify or clone this [cython helper](https://github.com/pydata/pandas/blob/master/pandas/tslib.pyx#L4889) to handle your case, but definitely not necessary, what you're doing is pretty quick already.

jreback commented on pandas/tseries/tests/test_frequencies.py:
```diff
@@ -52,6 +52,16 @@ def test_to_offset_multiple():
+    assert (result == expected)
+
+    freqstr = '2SM-16'
+    result = frequencies.to_offset(freqstr)
```
add some tests where this fails (e.g. it *should* fail)

jreback replied:
add ``SM-15`` as well (which should work)

jreback commented on pandas/tseries/tests/test_frequencies.py:
```diff
@@ -137,6 +151,16 @@ def test_anchored_shortcuts():
     assert (result1 == expected)
 
+    result1 = frequencies.to_offset('SM')
+    result2 = frequencies.to_offset('SM-15')
```
you did, nvm

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -51,6 +51,53 @@ New behaviour:
+
+Pandas has gained new frequency offsets, ``SemiMonthEnd`` ('SM') and ``SemiMonthBegin`` ('SMS').
+These provide date offsets anchored (by default) to the 15th and end of month, and 15th and 1st of month respectively.
+Optionally you can specify the ``day_of_month`` which will be used instead of the 15th each month. Anchoring suffixes
```
The last 2 sentences are not necessary, the doc-string and what you are doing below covers that.

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -51,6 +51,53 @@ New behaviour:
+
+.. ipython:: python
+
+    Timestamp('2016-01-01') + SemiMonthBegin(day_of_month=16)
```
No need to show this, it is essentially an implementation detail (the ``day_of_month`` one), the anchor is much more important.

adrienemery commented:
- updated whatsnew
- added more tests (including failing cases)

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -51,6 +51,43 @@ New behaviour:
+
+    pd.date_range('2015-01-01', freq='SMS', periods=4)
+
+Using the anchoring suffix:
```
Using the anchoring suffix, you can also specify the day of month

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+
+    @classmethod
+    def _from_name(cls, suffix=None):
+        if not suffix:
```
day_of_month = int(suffix or cls._default_day_of_month)

adrienemery replied:
`._from_name`  is used here: https://github.com/pydata/pandas/blob/master/pandas/tseries/frequencies.py#L566 

without it you get a value error when calling `.to_offset`
```bash
In [4]: pd.tseries.frequencies.to_offset('SMS')
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/Users/adrienemery/Code/pandas/pandas/tseries/frequencies.py in to_offset(freqstr)
    464             for stride, name, _ in opattern.findall(freqstr):
--> 465                 offset = get_offset(name)
    466                 if stride_sign is None:

/Users/adrienemery/Code/pandas/pandas/tseries/frequencies.py in get_offset(name)
    565             # many '-')
--> 566             offset = klass._from_name(*split[1:])
    567         except (ValueError, TypeError, KeyError):

AttributeError: type object 'SemiMonthBegin' has no attribute '_from_name'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-4-add2ac572fb8> in <module>()
----> 1 pd.tseries.frequencies.to_offset('SMS')

/Users/adrienemery/Code/pandas/pandas/tseries/frequencies.py in to_offset(freqstr)
    475                     delta = delta + offset
    476         except Exception:
--> 477             raise ValueError("Could not evaluate %s" % freqstr)
    478
    479     if delta is None:

ValueError: Could not evaluate SMS
```

jreback replied:
oh ok, so continue to have that, just do the init in the constructor as you have now, no need to have same code in 2 places

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+
+    @property
+    def rule_code(self):
+        suffix = '-%s' % self.day_of_month
```
use ``.format()``

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+                n += 1
+
+        months = n // 2
+        day = 31 if n % 2 else self.day_of_month
```
do you have a tests that goes thru all of the months in a single year and verifies correctness? what about leap year for feb?

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+        else:
+            roll = np.where(after_day_of_month, n + 2, n + 1)
+
+        return ((i.to_period('M') + (roll // 2)).to_timestamp() +
```
I would like to see this split up a bit, a lot going on here, can you make it more clear?

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+            roll_start = np.where(is_month_start, -1, 0)
+            roll = roll_after + roll_start
+
+        return ((i.to_period('M') + (roll // 2)).to_timestamp() +
```
same  here

jreback commented on pandas/tseries/tests/test_frequencies.py:
```diff
@@ -137,6 +165,39 @@ def test_anchored_shortcuts():
+
+    # ensure invalid cases fail as expected
+    invalid_anchors = ['SM-0', 'SM-28', 'SM-29',
+                       'SMS-1', 'SMS-28', 'SMS-30']
```
add non-numeric after the anchor ``SMS-FOO`` for invalid tests

jreback commented:
@adrienemery looks real good. just some minor comments. ping when pushed and green.

adrienemery commented:
@jreback should be good to go - I have addressed comments and build is green

jreback commented:
@adrienemery thanks! nice PR. these have a lot of pieces for a full patch. Nice job!

give a check on the built docs (prob take a few hours), and if needs corrections, pls issue a followup PR.

adrienemery commented:
@jreback thanks for your help on this PR - this is my first time contributing to pandas. Also thanks for all your work on pandas 👍
